manage-data/ingest/transform-enrich/ingest-lag.md

```diff
@@ -0,0 +1,312 @@
+
+## Architectures
+
+Regardless of the chosen architecture, it's a good practice to add a `remove` processor at the end of the pipeline to drop the `_tmp` field. The raw timestamps from the various processing steps are not needed, as the latency in seconds should be sufficient. For additional pipeline architectures, refer to the [Ingest Architectures](../ingest-reference-architectures.md) documentation.
+
+## Logstash
+
+When Logstash is the mix we want to add a timestamp, this can only be done by using Ruby and the simplest form is this:
+
+```logstash
+ruby {
+  code => "event.set('[_tmp][logstash_seen]', Time.now());"
+}
+```
+
+### Elastic Agent => Elasticsearch
+
+We can use `@timestamp` and `event.ingested` and calculate the difference. This will give you the following document. The `event.ingestion.latency` is in seconds.
+
+```json
+{
+  "event": {
+    "ingestion": {
+      "latency": 443394
+    }
+  }
+}
+```
+
+#### Script
+
+```json
+POST _ingest/pipeline/_simulate
+{
+    "docs": [{
+        "_source": {
+            "@timestamp": "2025-04-03T10:00:00.000Z",
+            "message": "user: philipp has logged in",
+            "_tmp": {
+              "logstash": "2025-04-03T10:00:02.456Z"
+            }
+
+        }
+    }],
+    "pipeline": {
+        "processors": [
+          {
+            "script": {
+              "description": "Calculates entire ingestion flow latency",
+              "if": "ctx['@timestamp'] != null",
+              "source": """
+                ZonedDateTime start = ZonedDateTime.parse(ctx['@timestamp']);
+                ctx.putIfAbsent("event", [:]);
+                ctx.event.putIfAbsent("ingestion", [:]);
+                ctx.event.ingestion.latency= ChronoUnit.SECONDS.between(start, metadata().now);
+              """
+            }
+          }
+        ]
+    }
+}
+```
+
+### Elastic Agent => Logstash => Elasticsearch
+
+In this scenario, we have an additional hop to manage. Elastic Agent populates the `@timestamp`, but Logstash does not add any timestamp by default. We recommend adding a temporary timestamp, for example by setting `_tmp.logstash_seen`. With this, you can calculate the following latency values:
+
+- Total latency: (`@timestamp - event.ingested`)
+- Elastic Agent => Logstash: (`@timestamp - _tmp.logstash_seen`)
+- Logstash => Elasticsearch: (`_tmp.logstash_seen - event.ingested`)
+
+These values can be especially helpful for debugging, as they allow you to quickly determine where the lag is introduced. Is the delay caused by the transfer from Elastic Agent to Logstash, or from Logstash to Elasticsearch?
+
+Below is a script that calculates these differences, providing latency values for each of the stages mentioned above.
+
+```json
+{
+  "event": {
+    "ingestion": {
+      "latency_logstash_to_elasticsearch": 443091,
+      "latency": 443093,
+      "latency_elastic_agent_to_logstash": 1
+    }
+  }
+}
+```
+
+#### Script
```

Comment:
Example script? What is this script for? Rewrite the heading to something more descriptive.